os/drivers/compression/compress.c

```diff
@@ -0,0 +1,106 @@
+/****************************************************************************
+ *
+ * Copyright 2023 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+
+#include <tinyara/fs/fs.h>
+#include <tinyara/fs/ioctl.h>
+#include <tinyara/compression.h>
+
+#include <errno.h>
+#include <debug.h>
+#include <assert.h>
+
+/****************************************************************************
+ * Function Prototypes
+ ****************************************************************************/
+
+static int comp_ioctl(FAR struct file *filep, int cmd, struct compress_header *arg);
+static ssize_t comp_read(FAR struct file *filep, FAR char *buffer, size_t len);
+static ssize_t comp_write(FAR struct file *filep, FAR const char *buffer, size_t len);
+
+/****************************************************************************
+ * Data
+ ****************************************************************************/
+static const struct file_operations compress_fops = {
+	0,                          /* open */
+	0,                          /* close */
+	comp_read,                  /* read */
+	comp_write,                 /* write */
+	0,                          /* seek */
+	comp_ioctl              /* ioctl */
+#ifndef CONFIG_DISABLE_POLL
+	, 0                         /* poll */
+#endif
+};
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static ssize_t comp_read(FAR struct file *filep, FAR char *buffer, size_t len)
+{
+	return 0;
+}
+
+static ssize_t comp_write(FAR struct file *filep, FAR const char *buffer, size_t len)
+{
+	return 0;
+}
+
+
+/************************************************************************************
+ * Name: comp_ioctl
+ *
+ * Description: The ioctl method for compress.
+ *
+ ************************************************************************************/
+static int comp_ioctl(FAR struct file *filep, int cmd, struct compress_header *arg)
```

Comment:
instead of 'struct compress_header', you should keep unsigned long arg.
Please check defined ioctl in fs.h

Reply:
I need a struct compress_header argument to apply compression on the struct data . The data to which compression or decompression will be applied , will be passed to ioctl as this arg

Reply:
Please check other drivers. They use `unsigned long` and do typecasting internally because of prototype of ioctl. They don't use proper type directly as you did. That's what Taejun said.